### PR TITLE
PrePostProcessor.output() - first implementation of post-processing

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - Supports convert_layout, convert_element_type and custom operations
 - Support of implicit conversions to user's tensor element type and layout
 - Introduced 'get_layout' and 'set_layout' for Result
 - Test coverage - 100%

### Tickets:
 - 62178

### CODE REVIEW

File: ngraph/test/preprocess.cpp [hunk 1]
```diff
@@ -327,3 +330,208 @@ TEST(pre_post_process, resize_no_tensor_width) {
...
+// --- PostProcess - set/convert layout ---
+TEST(pre_post_process, postprocess_set_layout_network) {
+    auto f = create_simple_function(element::f32, Shape{1, 3, 2, 2});
+    f = PrePostProcessor().output(OutputInfo().network(OutputNetworkInfo().set_layout("NCHW"))).build(f);
+    EXPECT_EQ(f->get_results()[0]->get_layout(), "NCHW");
+}
+
+TEST(pre_post_process, postprocess_set_layout_tensor) {
+    auto f = create_simple_function(element::f32, Shape{1, 3, 2, 2});
+    // no layout is specified for network, no way to implicitly convert it to user's layout
```
Comment (ilya-lavrenov) [inline ngraph/test/preprocess.cpp:397]: will we have the same behavior for inputs, right?

File: ngraph/test/preprocess.cpp [hunk 2]
```diff
@@ -327,3 +330,208 @@ TEST(pre_post_process, resize_no_tensor_width) {
...
+}
+
+TEST(pre_post_process, postprocess_implicit_convert_element_type_and_layout) {
+    auto f = create_simple_function(element::f32, Shape{1, 3, 2, 2});
+    f = PrePostProcessor()
+            .output(OutputInfo()
+                        .network(OutputNetworkInfo().set_layout("NCHW"))
+                        .tensor(OutputTensorInfo().set_layout("NHWC").set_element_type(element::u8)))
+            .build(f);
+    EXPECT_EQ(f->get_results()[0]->get_element_type(), element::u8);
```
Comment (ilya-lavrenov) [inline ngraph/test/preprocess.cpp:490]: should not we check shape order as well?

File: ngraph/core/src/preprocess/preprocess_steps_impl.cpp [hunk 3]
```diff
@@ -166,5 +166,44 @@ void PreProcessSteps::PreProcessStepsImpl::add_convert_layout_impl(const Layout&
...
+//------------- Post processing ------
+void PostStepsList::add_convert_impl(const ov::element::Type& type) {
+    m_actions.emplace_back(std::make_tuple(
+        [type](const ov::Output<Node>& node, PostprocessingContext& ctxt) -> ov::Output<ov::Node> {
+            ov::element::Type t = type;
+            if (t == element::Type{}) {
+                t = ctxt.target_element_type();
+            }
+            if (t == node.get_node()->get_element_type()) {
+                return node;
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/core/src/preprocess/preprocess_steps_impl.cpp:178]: yes, will fix

File: ngraph/core/src/preprocess/pre_post_process.cpp [hunk 4]
```diff
@@ -271,6 +363,69 @@ std::shared_ptr<Function> PrePostProcessor::build(const std::shared_ptr<Function
...
+            new_result->set_layout(context.layout());
+        }
+        new_result->get_input_tensor(0).set_names(result->get_input_tensor(0).get_names());
+        function->add_results({new_result});
+        function->remove_result(result);
+    }
+
+    // Validate nodes
     if (tensor_data_updated) {
         function->validate_nodes_and_infer_types();
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/core/src/preprocess/pre_post_process.cpp:511]: Final validation after adding post-processing is needed anyway. I doubt that we need to call it twice (after pre-processing and after post-processing)

File: ngraph/test/preprocess.cpp (hunk 2 above)
Comment (nosovmik) [reply ngraph/test/preprocess.cpp:762]: Yes, missed this, thanks, fixes

**Review by ilya-lavrenov — APPROVED** [review]

File: ngraph/core/src/preprocess/pre_post_process.cpp [hunk 5]
```diff
@@ -346,9 +444,73 @@ std::shared_ptr<Function> PrePostProcessor::build(const std::shared_ptr<Function
...
+            auto action_result = action({node}, context);
+            node = std::get<0>(action_result);
+        }
+
+        // Create result
+        auto new_result = std::make_shared<ov::op::v0::Result>(node);
+        if (!context.layout().empty()) {
+            new_result->set_layout(context.layout());
+        }
+        new_result->get_input_tensor(0).set_names(result->get_input_tensor(0).get_names());
```
Comment (ilya-lavrenov) [inline ngraph/core/src/preprocess/pre_post_process.cpp:509]: should not we reset tensor names for the previous / original tensor?

How is it done for inputs?

**Pull request merged**